### ISSUE TITLE
Clippy fails to auto-fix `search_is_some` issue when `.then_some()` is used

### ISSUE THREAD

**Issue by andrewswait** [rust-lang/rust-clippy#11910]
### Summary

When `clippy` attempted to auto-fix an instance of [`search_is_some`](https://rust-lang.github.io/rust-clippy/master/index.html#search_is_some), there was a compiler error. 

The use of `.any()` is followed by `.then_some(...)` so it needs to add parentheses around the part it's negating with `!` rather than attempting to negate the entire expression (you can't 'negate' `Option<T>`).

### Reproducer

I tried this code:

```rust
impl Game {
    pub fn is_valid(&self, max_red: u32, max_green: u32, max_blue: u32) -> Option<u32> {
        self.turns
            .iter()
            .find(|t| !t.is_valid(max_green, max_blue, max_red))
            .is_none()
            .then_some(self.id)
    }
}
```

I expected to see this happen:

```rust
impl Game {
    pub fn is_valid(&self, max_red: u32, max_green: u32, max_blue: u32) -> Option<u32> {
        (!self
            .turns
            .iter()
            .any(|t| !t.is_valid(max_green, max_blue, max_red)))
        .then_some(self.id)
    }
}
```

Instead, this happened:

```
The following errors were reported:
error[E0600]: cannot apply unary operator `!` to type `std::option::Option<u32>`
  --> src/day2/part1.rs:80:9
   |
80 | /         !self.turns
81 | |             .iter().any(|t| !t.is_valid(max_green, max_blue, max_red))
82 | |             .then_some(self.id)
   | |_______________________________^ cannot apply unary operator `!`

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0600`.
```

### Version

```text
rustc 1.76.0-nightly (8c2b57721 2023-12-01)
binary: rustc
commit-hash: 8c2b57721728233e074db69d93517614de338055
commit-date: 2023-12-01
host: x86_64-apple-darwin
release: 1.76.0-nightly
LLVM version: 17.0.5
```

### Additional Labels

@rustbot labels +I-suggestion-causes-error
